collector/lib/ConfigLoader.cpp

```diff
@@ -8,114 +14,520 @@ namespace collector {
+      return {{"Unsupport repeated type MESSAGE"}};
+    } break;
+
+    default: {
+      ParserError err;
+      err << "Unknown type " << field->type_name();
+      return {{err}};
+    }
+  }
+}
+
+ParserError ParserYaml::WrapError(const std::exception& e) {
+  ParserError err;
+  err << file_ << ": " << e.what();
+  return err;
+}
+
+template <typename T>
+std::variant<T, ParserError> ParserYaml::TryConvert(const YAML::Node& node) {
+  try {
+    return node.as<T>();
+  } catch (YAML::InvalidNode& e) {
+    return WrapError(e);
+  } catch (YAML::BadConversion& e) {
+    return WrapError(e);
   }
+}
+
+std::string ParserYaml::SnakeCaseToCamel(const std::string& s) {
+  std::string out;
+  bool capitalize = false;
 
-  sensor::ExternalIpsEnabled enable_external_ips;
-  std::string enabled_value = external_ips_node["enabled"] ? external_ips_node["enabled"].as<std::string>() : "";
-  std::transform(enabled_value.begin(), enabled_value.end(), enabled_value.begin(), ::tolower);
+  for (const auto& c : s) {
+    if (c == '_') {
+      capitalize = true;
+      continue;
+    }
+
+    if (capitalize) {
+      out += (char)std::toupper(c);
+    } else {
+      out += c;
+    }
+    capitalize = false;
+  }
+
+  return out;
+}
+
+std::string ParserYaml::CamelCaseToSnake(const std::string& s) {
+  std::string out;
+  bool first = true;
+
+  for (const auto& c : s) {
+    if (!first && std::isupper(c) != 0) {
+      out += '_';
+    }
+    out += (char)std::tolower(c);
+    first = false;
+  }
+
+  return out;
+}
+
+ConfigLoader::ConfigLoader(CollectorConfig& config)
+    : config_(config), parser_(CONFIG_FILE.value()) {}
+
+void ConfigLoader::Start() {
+  thread_.Start([this] { WatchFile(); });
+  CLOG(INFO) << "Watching configuration file: " << parser_.GetFile().string();
+}
+
+void ConfigLoader::Stop() {
+  thread_.Stop();
+  CLOG(INFO) << "No longer watching configuration file: " << parser_.GetFile().string();
+}
+
+ConfigLoader::Result ConfigLoader::LoadConfiguration(const std::optional<const YAML::Node>& node) {
+  sensor::CollectorConfig runtime_config = NewRuntimeConfig();
+  ParserResult errors;
+
+  if (!node.has_value()) {
+    if (!stdf::exists(parser_.GetFile())) {
+      return FILE_NOT_FOUND;
+    }
+    errors = parser_.Parse(&runtime_config);
 
-  if (enabled_value == "enabled") {
-    enable_external_ips = sensor::ExternalIpsEnabled::ENABLED;
-  } else if (enabled_value == "disabled") {
-    enable_external_ips = sensor::ExternalIpsEnabled::DISABLED;
   } else {
-    CLOG(WARNING) << "Unknown value for for networking.externalIps.enabled. Setting it to DISABLED";
-    enable_external_ips = sensor::ExternalIpsEnabled::DISABLED;
+    errors = parser_.Parse(&runtime_config, *node);
   }
-  int64_t max_connections_per_minute = networking_node["maxConnectionsPerMinute"].as<int64_t>(CollectorConfig::kMaxConnectionsPerMinute);
 
+  if (errors) {
+    CLOG(ERROR) << "Failed to parse " << parser_.GetFile();
```

Comment:
Since collector has multiple threads it is possible that this line will be logged, then other unrelated logging statements will be created, and then the errors in the loop will be logged, again possibly interspersed by unrelated logging statements. I think it would be best to save the error messages to one string and then log that string.

Reply:
I wonder if there's any benefit to officially supporting that in the logging library (presumably it can be done by adding a specialisation of the `<<` operator?)

Reply:
I'll look into it, the logging library is purposefully unsafe w.r.t threads because you don't want to be blocking proper operations with log messages. I'm also not a big fan of copying a bunch of strings around, so in order for me to be onboard with this, I'll need to figure out whether there's a way to do it by moving strings around, but there probably isn't.